# Hacker News Posts
we'll work with a dataset of submissions to popular technology site [Hacker News.]('https://news.ycombinator.com/' 'hacker news')
Hacker News is a site started by the startup incubator [Y Combinator]('https://www.ycombinator.com/'), where user-submitted stories (known as "posts") receive votes and comments, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of the Hacker News listings can get hundreds of thousands of visitors as a result.

You can find the data set [here]('https://www.kaggle.com/hacker-news/hacker-news-posts'), but note that we have reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that didn't receive any comments and then randomly sampling from the remaining submissions. Below are descriptions of the columns:
* `id`: the unique identifier from Hacker News for the post
* `title`: the title of the post
* `url`: the URL that the posts links to, if the post has a URL
* `num_points`: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* `num_comments`: the number of comments on the post
* `author`: the username of the person who submitted the post
* `created_at`: the date and time of the post's submission
We're mostly interested in the posts that have titles which comprise `Ask HN` OR `Show HN`. Normally, Users submit `Ask HN` posts to ask the Hacker News community a specific question.
same to `Show HN` where users submit `Show HN` posts to show the Hacker News community a project, product, or just something interesting.
we will compare these posts according to the two properties to determine:
* Do `Ask HN` or `Show HN` receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [2]:
from csv import reader as rd

In [3]:
open_file = open('hacker_news.csv')
read_file = rd(open_file)
hn = list(read_file)

In [4]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Above, we've read our data and put it into a list of lists. We went ahead to display the first five rows including the header.
Now we will have to remove the header to be able to analyze our data.

In [5]:
headers = hn[0]
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [6]:
# displaying the first five rows
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now we have to find the posts that start with `Ask HN` or `Show HN` whereby we will us the ***string method*** called `startswith`. i.e

```python
    print('dataquest'.startswith('Data'))
    print('dataquest'.startswith('data'))
```
<code>
    False
    True
</code>
we get `False` on the first try because `dataquest` doesn't start with `Data`. Whereas, the second will print `True`. ***Capitalization also matters***.
Maybe we would wish to control the case, we can use the `lower` method which returns a lowercase version of the string at the beginning. i.e

```python
print('DataQuest'.lower())
```

```
dataquest
```

Now let's seperate the posts based on having `Ask HN` and `Show HN` at the beginning which we will put them in different lists

In [7]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    #print(title)
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print("Ask posts ==> ", len(ask_posts))
print("Show posts ==> ", len(show_posts))
print("Other posts ==> ", len(other_posts))

Ask posts ==>  1744
Show posts ==>  1162
Other posts ==>  17194


From the above result, we can see that `Ask HN` posts are ***1744*** in total, `Show HN` posts are ***1162*** in total and the `other posts` are ***17194*** in total.
Now we we are going to calculate the average number of comments for our main two lists for `ask` and `show` posts to determine which one has more posts on average

In [8]:
# the first five rows of ask posts
print("Ask Posts\n\n", ask_posts[:5])

Ask Posts

 [['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [9]:
# the first five rows of show posts
print("Show Posts\n\n", show_posts[:5])

Show Posts

 [['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


In [10]:
total_ask_comments = 0
for row in ask_posts:
    comments = int(row[4])
    # print(comments, "\n")
    total_ask_comments += comments
# print(type(comments))
avg_ask_comments = total_ask_comments / len(ask_posts) 
print("Ask comments total ==> ", total_ask_comments)
print("Average ask comments ==> ", avg_ask_comments)

Ask comments total ==>  24483
Average ask comments ==>  14.038417431192661


In [11]:
# for show posts
total_show_comments = 0
for row in show_posts:
    comments = int(row[4])
    total_show_comments += comments
avg_show_comments = total_show_comments / len(show_posts)
print("Total show comments ==> ", total_show_comments)
print("Average show comments ==> ", avg_show_comments)

Total show comments ==>  11988
Average show comments ==>  10.31669535283993


From the above result we can see that `ask posts` receive more comments on average compared to `show posts`.

***Ask posts*** receive approximately `14` comments on average whereas, ***Show posts*** receive approximately `10` comments on average.

Now we will put our focus on ***ask posts*** because they receive a higher number of comments on average.

Now we will check if `ask posts` created at a certain time are more likely to attract comments by:
- Calculating the number of posts created in each hour of the day, along with the number of comments received.
- calculate the average number of comments received by hour created.

We will first determine the number of posts created in each hour of the day, along with the number of comments received. We will use `datetime` to work with data in the `created_at` column.

We will use `datetime.strptime()` to ***parse*** the dates which are stored as strings ad return ___datetime___ objects to work with.

```python
    date_1_str = "December 24, 1984"
    date_1_dt = dt.datetime.strptime(date_1_str, "%B %d, %Y")
```

where, `%B` is full name of ***month***, `%d` is a ***two-digit day*** and `%Y` is a ****four-digit year***.

Now let's calculate the number of ask posts per hour along with the total number of comments.

In [12]:
import datetime as dt

In [13]:
print(dir(dt))

['MAXYEAR', 'MINYEAR', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'date', 'datetime', 'datetime_CAPI', 'sys', 'time', 'timedelta', 'timezone', 'tzinfo']


In [18]:
result_list = []
for row in ask_posts:
    created_at = row[6]
    comments = int(row[4])
    result_list.append([created_at, comments])
# print(result_list)
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
#     if time in counts_by_hour:
#         comments_by_hour[time] += comment
#         counts_by_hour[time] += 1
#     else:
#         comments_by_hour[time] = comment
#         counts_by_hour[time] = 1
    if time not in counts_by_hour:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1
    else:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1

In [19]:
print("Counts per hr \n\n", counts_by_hour)

Counts per hr 

 {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


In [20]:
print("comments per hr \n\n", comments_by_hour)

comments per hr 

 {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


We've created two dictionaries which consist of:
* `counts_by_hour` - contains the number of ask posts created during each hour of the day.
* `comments_by_hour` - contains the corresponding number of comments ask posts created at each hour received.

Now we are going to use these dictionaries to calculate the average number of comments for the created posts during each and every hour of the day.

We are going to initialize an empty list to store average per hour.

the following is what we are going to do:
* Initialize an enpty list for the list of lists
* iterate over the dictionaries and append to the list of lists as lists of average number of comments and hours during which posts are created.

calculating the average number of comments per post for posts created during each hour of the day.

In [24]:
avg_by_hour = []
for com in comments_by_hour:
    avg_by_hour.append([com, comments_by_hour[com] / counts_by_hour[com]])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


We've calculated the average number of comments for posts created during each hour of the day, and stored the results in a list of lists named avg_by_hour.

Reading the hours seems challenging. We will sort the list of list so that we may be able to determine the hours with the highest values.

In [26]:
swap_avg_by_hour = []
for rows in avg_by_hour:
    swap_avg_by_hour.append([rows[1], rows[0]])
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [30]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [35]:
print( "Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[:5]:
    average = row[0]
    hr = row[1]
    time = dt.datetime.strptime(hr, "%H").strftime("%H:%M")
    top_5 = "{time}: {average:.2f} average comments per post".format(time=time, average=average)
    print(top_5)

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


From our findings, the hour that receives the most comments per post on average is `15:00`, with an average of `38.59` comments per post. There's about a `60%` increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation]('https://www.kaggle.com/hacker-news/hacker-news-posts' 'Documentation for the data set'), the timezone used is `Eastern Time` in the `US`. So, we could also write `15:00 as 3:00 pm est`.

But my timezone is `UTC+3` which means it's ***7 hrs** ahead of `Eastern Time`. So the best hour according to my timezone is `22:00 as 10:00 pm utc+3`.


# Conclusion
In the entire project, we analyzed ask posts and show posts to determine which type of post and time receives the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between `15:00/22:00` and `16:00/23:00` ___(3:00 pm est/10:00 pm utc+3 - 4:00 pm est/11:00 pm utc+3)___.

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.